internal/codeintel/syntactic_indexing/scheduler.go

```
@@ -0,0 +1,144 @@
+	batchOptions := reposcheduler.NewBatchOptions(schedulerConfig.RepositoryProcessDelay, true, &schedulerConfig.PolicyBatchSize, schedulerConfig.RepositoryBatchSize)
+
+	repos, err := s.RepositorySchedulingService.GetRepositoriesForIndexScan(ctx,
+		batchOptions, currentTime)
+
+	if err != nil {
+		return err
+	}
+
+	commitsToSchedule := make(map[api.RepoID]collections.Set[api.CommitID])
+	enqueueOptions := EnqueueOptions{force: false}
+
+	var allErrors errors.MultiError
+
+	for _, repoToIndex := range repos {
+		repo, _ := s.RepoStore.Get(ctx, api.RepoID(repoToIndex.ID))
+		policyIterator := internal.NewPolicyIterator(s.PoliciesService, repoToIndex.ID, internal.SyntacticIndexing, schedulerConfig.PolicyBatchSize)
+		err := policyIterator.ForEachPoliciesBatch(ctx, func(policies []policiesshared.ConfigurationPolicy) error {
+			commitMap, err := s.PolicyMatcher.CommitsDescribedByPolicy(ctx, int(repoToIndex.ID), repo.Name, policies, currentTime)
+
+			if err != nil {
+				return err
+			}
+
+			for commit, policyMatches := range commitMap {
+				if len(policyMatches) == 0 {
+					continue
+				}
+				if commits := commitsToSchedule[repo.ID]; commits != nil {
+					commits.Add(api.CommitID(commit))
+				} else {
+					commitsToSchedule[repo.ID] = collections.NewSet(api.CommitID(commit))
+				}
+			}
+
+			return nil
+		})
+
+		if err != nil {
+			allErrors = errors.Append(allErrors, errors.Newf("Failed to discover commits eligible for syntactic indexing for repo [%s]: %v", repo.Name, err))
+		}
+	}
+
+	for repoId, commits := range commitsToSchedule {
+		for _, commitId := range commits.Values() {
+			if _, err := s.Enqueuer.QueueIndexingJobs(ctx, repoId, commitId, enqueueOptions); err != nil {
+				allErrors = errors.Append(allErrors, errors.Newf("Failed to schedule syntactic indexing of repo [ID=%s], commit [%s]: %v", repoId, commitId, err))
+			}
+		}
+	}
```

@varungandhi-src I've re-done this part, main highlights:

1. Collecting all repos and commits once
2. Removing fail fast behavior in repo iterator
3. Removing fail fast behavioru in (repo, commit) iterator

Additionally the enqueuer no longer performs revision checks, we'll leave that to the syntactic worker itself.

This means the scheduler will always do best effort scheduling, while returning all accumulated errors.

I hope my understanding of `return allErrors` is right here and equivalent to `return nil` if no `allErrors = errors.Append` calls ever happened.

---

Thanks, this looks much better!

> I hope my understanding of return allErrors is right here and equivalent to return nil if no allErrors = errors.Append calls ever happened.

Yep, we're following this pattern elsewhere too. I think you don't even need to declare the type as `MultiError`, you can see other places where `errors.Append(` can take the first argument as just `error` instead of `MultiError`. (search for `var errs error`)